src/ffi/android.rs

```diff
@@ -201,19 +222,28 @@ pub unsafe extern "system" fn Java_org_c4dt_artiwrapper_TorLibApi_torRequest(
     headers_j: JObject,
     body_j: jbyteArray,
 ) -> jobject {
-    let (cache_dir, request) =
-        match make_request(env, cache_dir_j, method_j, url_j, headers_j, body_j) {
-            Ok(v) => v,
-            Err(e) => {
-                let _ = env.throw((
-                    "org/c4dt/artiwrapper/TorLibException",
-                    format!("make request: {:?}", e),
-                ));
-                return JObject::null().into_inner();
-            }
-        };
+    let request = match make_request(env, method_j, url_j, headers_j, body_j) {
+        Ok(v) => v,
+        Err(e) => {
+            let _ = env.throw((
+                "org/c4dt/artiwrapper/TorLibException",
+                format!("make request: {:?}", e),
+            ));
+            return JObject::null().into_inner();
+        }
+    };
```

Comment:
An exercice for the reader (hint: `unwrap_or_else`) :)

Reply:
But then I end up with `request` that contains a value in case of error, correct? How do I exit early then?

Reply:
Ho, you're right, I was a bit fast on that, it indeed returns a value. But you can work around it using the same trick you used in `Java_org_c4dt_artiwrapper_TorLibApi_hello`: wrap it the whole function into a closure returning a `Result`.

```rust
    || -> Result<JObject> {
        let request =
            make_request(env, method_j, url_j, headers_j, body_j).context("make request")?;
        trace!("Request: {:?}", request);

        let cache_dir = get_cache_dir(env, cache_dir_j).context("get cache dir")?;
        let client = Client::new(PathBuf::from(cache_dir));

        let response = client.send(request).context("send request")?;
        trace!("Response: {:?}", response);

        format_response(env, response).context("format response")
    }()
    .unwrap_or_else(|e| {
        let _ = env.throw(("org/c4dt/artiwrapper/TorLibException", e.to_string()));
        JObject::null()
    })
    .into_inner()
```

Reply:
Ooooh, yes indeed, beautiful! :heart: 

(I had to use a `format!("{:?}", e)` otherwise it seems the message would only contain the last context :thinking: )